### PR TITLE
Add example DAG to demostrate emitting approaches

### PR DESCRIPTION
**Pull request by uranusjr**

### CODE REVIEW

File: airflow/example_dags/example_dataset_event_extra.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+ds = Dataset("s3://output/1.txt")
+
+with DAG(
+    dag_id="produce_with_yield",
+    catchup=False,
+    start_date=datetime.datetime.min,
+    schedule="@daily",
+    tags=["produces", "dataset-scheduled"],
+) as dag:
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]:
```suggestion
):
```

File: airflow/example_dags/example_dataset_event_extra.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    produce()
+
+with DAG(
+    dag_id="produce_with_context",
+    catchup=False,
+    start_date=datetime.datetime.min,
+    schedule="@daily",
+    tags=["produces", "dataset-scheduled"],
+) as dag:
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]:
```suggestion
):
```

File: airflow/example_dags/example_dataset_event_extra.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+from airflow.operators.bash import BashOperator
+
+ds = Dataset("s3://output/1.txt")
+
+with DAG(
+    dag_id="produce_with_yield",
+    catchup=False,
+    start_date=datetime.datetime.min,
+    schedule="@daily",
+    tags=["produces", "dataset-scheduled"],
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]:
```suggestion
    tags=["produces"],
```

Not a dataset scheduled dag. Might be better with no tags?

File: airflow/example_dags/example_dataset_event_extra.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        yield Metadata(ds, {"hi": "bye"})
+
+    produce()
+
+with DAG(
+    dag_id="produce_with_context",
+    catchup=False,
+    start_date=datetime.datetime.min,
+    schedule="@daily",
+    tags=["produces", "dataset-scheduled"],
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]:
```suggestion
    tags=["produces"],
```

File: airflow/example_dags/example_dataset_event_extra.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+with DAG(
+    dag_id="produce_classic_from_execute",
+    catchup=False,
+    start_date=datetime.datetime.min,
+    schedule="@daily",
+    tags=["produces", "dataset-scheduled"],
+):
+
+    class EmittingBashOperator(BashOperator):
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]: I'm a little hesitant to show this approach in an example DAG. Feels too low level. If this is the right approach for folks, they probably don't need the example in the first place?

File: airflow/example_dags/example_dataset_event_extra.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        dataset_events[ds].extra = {"hi": "bye"}
+
+    produce()
+
+with DAG(
+    dag_id="produce_classic_from_execute",
+    catchup=False,
+    start_date=datetime.datetime.min,
+    schedule="@daily",
+    tags=["produces", "dataset-scheduled"],
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]:
```suggestion
    tags=["produces"],
```

File: airflow/example_dags/example_dataset_event_extra.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            return result
+
+    EmittingBashOperator(task_id="produce", outlets=[ds], bash_command=":")
+
+with DAG(
+    dag_id="produce_classic_from_task_hook",
+    catchup=False,
+    start_date=datetime.datetime.min,
+    schedule="@daily",
+    tags=["produces", "dataset-scheduled"],
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]:
```suggestion
    tags=["produces"],
```

**Review by jedcunningham — APPROVED** [review]

File: airflow/example_dags/example_dataset_event_extra.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+from airflow.datasets import Dataset
+from airflow.datasets.metadata import Metadata
+from airflow.models.dag import DAG
+from airflow.operators.bash import BashOperator
+
+ds = Dataset("s3://output/1.txt")
+
+with DAG(
+    dag_id="produce_with_yield",
```
Comment (jedcunningham) [inline airflow/example_dags/example_dataset_event_extra.py]: I wonder if "produce" is enough context for the dag_id. Maybe this instead, better matches the existing example dag_ids?

```suggestion
    dag_id="dataset_produces_with_yield",
```

Comment (uranusjr) [reply airflow/example_dags/example_dataset_event_extra.py]: *Dataset produces* seems a bit weird too… We don’t need to include *produce* in the ID at all. I’ll change this to something more descriptive (and longer).

**Comment by uranusjr** [comment]
This seems to have accidentally caught a bug in serialisation tests (not in Airflow itself but the test).

**Pull request merged**